src-tool/Pact/Analyze/Eval/Core.hs

```diff
@@ -231,6 +245,60 @@ evalCore (StrContains needle haystack) = do
     _sSbv (coerceS @Str @String needle')
     `SBVS.isInfixOf`
     _sSbv (coerceS @Str @String haystack')
+
+-- hash values
+-- TODO (RS): we should add warnings to the stack, allowing
+--            to return shims in case, the content is not statically known
+evalCore (StrHash sT) = eval sT <&> unliteralS >>= \case
+  Nothing -> throwErrorNoLoc notStaticErr
+  Just (Str b)  -> pure $ literalS . Str . T.unpack . asString $ pactHash (encodeUtf8 (T.pack b))
+
+evalCore (IntHash iT) = eval iT <&> unliteralS >>= \case
+  Nothing -> throwErrorNoLoc notStaticErr
+  Just i  -> pure $ literalS . Str . T.unpack . asString . pactHash
+    $ toStrict $ Aeson.encode $ Pact.PLiteral $ Pact.LInteger i
+    
+evalCore (BoolHash bT) = eval bT <&> unliteralS >>= \case
+  Nothing -> throwErrorNoLoc notStaticErr
+  Just b  -> pure $ literalS . Str . T.unpack . asString . pactHash
+    $ toStrict $ Aeson.encode b
+    
+evalCore (DecHash d) = eval d <&> unliteralS >>= \case
+  Nothing -> throwErrorNoLoc notStaticErr
+  Just d' -> pure $ literalS . Str . T.unpack . asString . pactHash
+    $ toStrict $ Aeson.encode $ Pact.PLiteral $ Pact.LDecimal (toPact decimalIso d')
+
+evalCore (ListHash ty' xs) = case ty' of
+  SStr -> do
+    S _ xs' <- eval xs
```

Comment:
This can be factored out into a helper function that takes a function argument `f` such that you use `pure (f <$> xs'')`.

Reply:
you mean like
```
 let wrap f = do
        S _ xs' <- eval xs
        case unliteral xs' of
          Nothing -> throwErrorNoLoc notStaticErr
          Just xs'' -> pure (f <$> xs'')
```
This does not typecheck because of (I think) there is no witness of what `ty'` is.